File: android/GrabDrop/app/src/main/java/com/grabdrop/network/NetworkManager.kt
// /GrabDrop/app/src/main/java/com/grabdrop/network/NetworkManager.kt
package com.grabdrop.network

import android.content.Context
import android.net.wifi.WifiManager
import android.os.Build
import android.util.Log
import com.grabdrop.util.Constants
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import org.json.JSONObject
import java.io.*
import java.net.*
import java.util.concurrent.ConcurrentHashMap

class NetworkManager(private val context: Context) {

    companion object {
        private const val TAG = "NetworkManager"
        private const val HEARTBEAT_INTERVAL = 3_000L
        private const val DEVICE_TIMEOUT = 10_000L
        private const val CLEANUP_INTERVAL = 5_000L

        private const val MSG_TYPE_HEARTBEAT = "HEARTBEAT"
        private const val MSG_TYPE_SCREENSHOT = "SCREENSHOT_READY"
    }

    private val scope = CoroutineScope(
        Dispatchers.IO + SupervisorJob() +
                CoroutineExceptionHandler { _, throwable ->
                    Log.e(TAG, "Coroutine error", throwable)
                }
    )

    private val _incomingOffers = MutableSharedFlow<ScreenshotOffer>(extraBufferCapacity = 10)
    val incomingOffers: SharedFlow<ScreenshotOffer> = _incomingOffers

    private val _nearbyDeviceCount = MutableStateFlow(0)
    val nearbyDeviceCount: StateFlow<Int> = _nearbyDeviceCount

    // deviceId -> last seen timestamp
    private val discoveredDevices = ConcurrentHashMap<String, DeviceInfo>()

    private var multicastLock: WifiManager.MulticastLock? = null
    private var udpListenerJob: Job? = null
    private var heartbeatJob: Job? = null
    private var cleanupJob: Job? = null
    private var tcpServerSocket: ServerSocket? = null

    private var latestScreenshotData: ByteArray? = null

    data class DeviceInfo(
        val deviceId: String,
        val deviceName: String,
        val address: InetAddress,
        var lastSeen: Long
    )

    fun start() {
        acquireMulticastLock()
        startUdpListener()
        startHeartbeat()
        startDeviceCleanup()
        Log.d(TAG, "Started all network components")
    }

    fun stop() {
        heartbeatJob?.cancel()
        cleanupJob?.cancel()
        udpListenerJob?.cancel()
        tcpServerSocket?.close()
        releaseMulticastLock()
        discoveredDevices.clear()
        _nearbyDeviceCount.value = 0
        scope.cancel()
        Log.d(TAG, "Stopped")
    }

    // --- Heartbeat ---

    private fun startHeartbeat() {
        heartbeatJob = scope.launch {
            // Send first heartbeat immediately
            sendHeartbeat()

            while (isActive) {
                delay(HEARTBEAT_INTERVAL)
                sendHeartbeat()
            }
        }
        Log.d(TAG, "Heartbeat started (interval=${HEARTBEAT_INTERVAL}ms)")
    }

    private suspend fun sendHeartbeat() {
        try {
            val json = JSONObject().apply {
                put("type", MSG_TYPE_HEARTBEAT)
                put("sender_id", Constants.DEVICE_ID)
                put("sender_name", Build.MODEL)
                put("timestamp", System.currentTimeMillis())
            }

            val data = json.toString().toByteArray()

            // Send to multicast group
            try {
                val address = InetAddress.getByName(Constants.MULTICAST_GROUP)
                val packet = DatagramPacket(data, data.size, address, Constants.UDP_PORT)
                DatagramSocket().use { socket ->
                    socket.broadcast = true
                    socket.send(packet)
                }
            } catch (e: Exception) {
                Log.e(TAG, "Multicast heartbeat failed", e)
            }

            // Also send to broadcast address
            try {
                val broadcastAddr = InetAddress.getByName("255.255.255.255")
                val packet = DatagramPacket(data, data.size, broadcastAddr, Constants.UDP_PORT)
                DatagramSocket().use { socket ->
                    socket.broadcast = true
                    socket.send(packet)
                }
            } catch (e: Exception) {
                Log.e(TAG, "Broadcast heartbeat failed", e)
            }
        } catch (e: Exception) {
            Log.e(TAG, "sendHeartbeat error", e)
        }
    }

    // --- Device Cleanup ---

    private fun startDeviceCleanup() {
        cleanupJob = scope.launch {
            while (isActive) {
                delay(CLEANUP_INTERVAL)
                cleanupStaleDevices()
            }
        }
    }

    private fun cleanupStaleDevices() {
        val now = System.currentTimeMillis()
        val stale = discoveredDevices.entries.filter { (_, info) ->
            now - info.lastSeen > DEVICE_TIMEOUT
        }
        for ((id, info) in stale) {
            discoveredDevices.remove(id)
            Log.d(TAG, "Device expired: ${info.deviceName} ($id)")
        }
        updateDeviceCount()
    }

    private fun updateDeviceCount() {
        val count = discoveredDevices.size
        if (_nearbyDeviceCount.value != count) {
            _nearbyDeviceCount.value = count
            Log.d(TAG, "Nearby devices: $count")
        }
    }

    // --- Screenshot Broadcast ---

    fun broadcastScreenshotAvailable(screenshotData: ByteArray) {
        latestScreenshotData = screenshotData
        scope.launch {
            val tcpPort = startTcpServer(screenshotData)
            sendScreenshotBroadcast(tcpPort, screenshotData.size)
        }
    }

    private fun sendScreenshotBroadcast(tcpPort: Int, fileSize: Int) {
        scope.launch {
            try {
                val json = JSONObject().apply {
                    put("type", MSG_TYPE_SCREENSHOT)
                    put("sender_id", Constants.DEVICE_ID)
                    put("sender_name", Build.MODEL)
                    put("tcp_port", tcpPort)
                    put("file_size", fileSize)
                    put("timestamp", System.currentTimeMillis())
                }

                val data = json.toString().toByteArray()

                try {
                    val address = InetAddress.getByName(Constants.MULTICAST_GROUP)
                    val packet = DatagramPacket(data, data.size, address, Constants.UDP_PORT)
                    DatagramSocket().use { socket ->
                        socket.broadcast = true
                        socket.send(packet)
                    }
                } catch (_: Exception) {}

                try {
                    val broadcastAddr = InetAddress.getByName("255.255.255.255")
                    val packet = DatagramPacket(data, data.size, broadcastAddr, Constants.UDP_PORT)
                    DatagramSocket().use { socket ->
                        socket.broadcast = true
                        socket.send(packet)
                    }
                } catch (_: Exception) {}

                Log.d(TAG, "Screenshot broadcast sent on port $tcpPort")
            } catch (e: Exception) {
                Log.e(TAG, "Failed to send screenshot broadcast", e)
            }
        }
    }

    // --- Download ---

    suspend fun downloadScreenshot(offer: ScreenshotOffer): ByteArray? =
        withContext(Dispatchers.IO) {
            try {
                Log.d(TAG, "Connecting to ${offer.senderAddress}:${offer.tcpPort}")
                val socket = Socket()
                socket.connect(
                    InetSocketAddress(offer.senderAddress, offer.tcpPort),
                    5000
                )

                val output = socket.getOutputStream()
                output.write("GET\n".toByteArray())
                output.flush()

                val input = socket.getInputStream()
                val dis = DataInputStream(input)
                val length = dis.readInt()
                Log.d(TAG, "Receiving $length bytes")

                val data = ByteArray(length)
                dis.readFully(data)

                socket.close()
                Log.d(TAG, "Download complete")
                data
            } catch (e: Exception) {
                Log.e(TAG, "Download failed", e)
                null
            }
        }

    // --- UDP Listener ---

    private fun startUdpListener() {
        udpListenerJob = scope.launch {
            // Try multicast first, fall back to plain broadcast
            try {
                startMulticastListener()
            } catch (e: Exception) {
                Log.e(TAG, "Multicast listener failed, trying fallback", e)
                startBroadcastListener()
            }
        }
    }

    private suspend fun startMulticastListener() {
        val group = InetAddress.getByName(Constants.MULTICAST_GROUP)
        val socket = MulticastSocket(Constants.UDP_PORT).apply {
            reuseAddress = true
            joinGroup(group)
            soTimeout = 0
        }

        val buffer = ByteArray(4096)
        Log.d(TAG, "Multicast UDP listener started on port ${Constants.UDP_PORT}")

        try {
            while (scope.isActive) {
                try {
                    val packet = DatagramPacket(buffer, buffer.size)
                    socket.receive(packet)
                    val message = String(packet.data, 0, packet.length)
                    handleUdpMessage(message, packet.address)
                } catch (_: SocketTimeoutException) {
                    // continue
                } catch (e: Exception) {
                    if (scope.isActive) {
                        Log.e(TAG, "UDP receive error", e)
                        delay(1000)
                    }
                }
            }
        } finally {
            try {
                socket.leaveGroup(group)
                socket.close()
            } catch (_: Exception) {}
        }
    }

    private suspend fun startBroadcastListener() {
        val socket = DatagramSocket(null).apply {
            reuseAddress = true
            bind(InetSocketAddress(Constants.UDP_PORT))
            broadcast = true
            soTimeout = 0
        }

        val buffer = ByteArray(4096)
        Log.d(TAG, "Broadcast UDP fallback listener started")

        try {
            while (scope.isActive) {
                try {
                    val packet = DatagramPacket(buffer, buffer.size)
                    socket.receive(packet)
                    val message = String(packet.data, 0, packet.length)
                    handleUdpMessage(message, packet.address)
                } catch (_: SocketTimeoutException) {
                    // continue
                } catch (e: Exception) {
                    if (scope.isActive) {
                        Log.e(TAG, "Fallback receive error", e)
                        delay(1000)
                    }
                }
            }
        } finally {
            socket.close()
        }
    }

    private suspend fun handleUdpMessage(message: String, senderAddress: InetAddress) {
        try {
            val json = JSONObject(message)
            val type = json.getString("type")
            val senderId = json.getString("sender_id")

            // Ignore our own messages
            if (senderId == Constants.DEVICE_ID) return

            val senderName = json.optString("sender_name", "Unknown")
            val timestamp = json.optLong("timestamp", System.currentTimeMillis())

            when (type) {
                MSG_TYPE_HEARTBEAT -> {
                    handleHeartbeat(senderId, senderName, senderAddress, timestamp)
                }

                MSG_TYPE_SCREENSHOT -> {
                    // Also treat screenshot broadcast as a heartbeat
                    handleHeartbeat(senderId, senderName, senderAddress, timestamp)

                    val offer = ScreenshotOffer(
                        senderId = senderId,
                        senderName = senderName,
                        senderAddress = senderAddress,
                        tcpPort = json.getInt("tcp_port"),
                        fileSize = json.getInt("file_size"),
                        timestamp = timestamp
                    )
                    Log.d(TAG, "Screenshot offer from $senderName (${senderAddress}:${offer.tcpPort})")
                    _incomingOffers.emit(offer)
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse UDP message", e)
        }
    }

    private fun handleHeartbeat(
        senderId: String,
        senderName: String,
        address: InetAddress,
        timestamp: Long
    ) {
        val existing = discoveredDevices[senderId]
        if (existing == null) {
            Log.d(TAG, "New device discovered: $senderName ($senderId) at $address")
        }

        discoveredDevices[senderId] = DeviceInfo(
            deviceId = senderId,
            deviceName = senderName,
            address = address,
            lastSeen = System.currentTimeMillis()
        )

        updateDeviceCount()
    }

    // --- TCP Server ---

    private suspend fun startTcpServer(data: ByteArray): Int = withContext(Dispatchers.IO) {
        tcpServerSocket?.close()

        val server = ServerSocket(0)
        tcpServerSocket = server
        val port = server.localPort

        Log.d(TAG, "TCP server started on port $port")

        scope.launch {
            try {
                server.soTimeout = Constants.SCREENSHOT_OFFER_TIMEOUT_MS.toInt() + 5000
                repeat(5) {
                    try {
                        val client = server.accept()
                        launch { handleTcpClient(client, data) }
                    } catch (_: SocketTimeoutException) {
                        return@launch
                    }
                }
            } catch (e: Exception) {
                if (e !is SocketException) {
                    Log.e(TAG, "TCP server error", e)
                }
            } finally {
                server.close()
            }
        }

        port
    }

    private suspend fun handleTcpClient(client: Socket, data: ByteArray) {
        withContext(Dispatchers.IO) {
            try {
                val reader = BufferedReader(InputStreamReader(client.getInputStream()))
                val request = reader.readLine()

                if (request == "GET") {
                    val dos = DataOutputStream(client.getOutputStream())
                    dos.writeInt(data.size)
                    dos.write(data)
                    dos.flush()
                    Log.d(TAG, "Sent ${data.size} bytes to ${client.inetAddress}")
                }
            } catch (e: Exception) {
                Log.e(TAG, "TCP client handler error", e)
            } finally {
                client.close()
            }
        }
    }

    // --- Multicast Lock ---

    private fun acquireMulticastLock() {
        try {
            val wifiManager = context.applicationContext
                .getSystemService(Context.WIFI_SERVICE) as WifiManager
            multicastLock = wifiManager.createMulticastLock("GrabDrop").apply {
                setReferenceCounted(true)
                acquire()
            }
            Log.d(TAG, "Multicast lock acquired")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to acquire multicast lock", e)
        }
    }

    private fun releaseMulticastLock() {
        multicastLock?.let {
            if (it.isHeld) it.release()
        }
        multicastLock = null
    }
}

File: desktop/network_manager.py
#!/usr/bin/python

#*================================================================
#*   Copyright (C) 2026 XUranus All rights reserved.
#*   
#*   File:         network_manager.py
#*   Author:       XUranus
#*   Date:         2026-03-14
#*   Description:  
#*
#================================================================*/

# /GrabDrop-Desktop/network_manager.py
import json
import logging
import socket
import struct
import threading
import time
from typing import Callable, Dict, Optional

import config

logger = logging.getLogger("Network")


class NetworkManager:
    def __init__(self):
        self.on_screenshot_offer: Optional[Callable[[dict], None]] = None

        self._running = False
        self._discovered_devices: Dict[str, dict] = {}
        self._devices_lock = threading.Lock()

        self._listener_thread: Optional[threading.Thread] = None
        self._heartbeat_thread: Optional[threading.Thread] = None
        self._cleanup_thread: Optional[threading.Thread] = None
        self._tcp_server_socket: Optional[socket.socket] = None

    @property
    def nearby_count(self) -> int:
        with self._devices_lock:
            return len(self._discovered_devices)

    def start(self):
        self._running = True
        self._listener_thread = threading.Thread(
            target=self._udp_listener, daemon=True
        )
        self._listener_thread.start()

        self._heartbeat_thread = threading.Thread(
            target=self._heartbeat_loop, daemon=True
        )
        self._heartbeat_thread.start()

        self._cleanup_thread = threading.Thread(
            target=self._cleanup_loop, daemon=True
        )
        self._cleanup_thread.start()

        logger.info("Network manager started")

    def stop(self):
        self._running = False
        if self._tcp_server_socket:
            try:
                self._tcp_server_socket.close()
            except Exception:
                pass
        logger.info("Network manager stopped")

    # --- Heartbeat ---

    def _heartbeat_loop(self):
        self._send_heartbeat()  # immediate first
        while self._running:
            time.sleep(config.HEARTBEAT_INTERVAL_S)
            if self._running:
                self._send_heartbeat()

    def _send_heartbeat(self):
        msg = json.dumps({
            "type": config.HEARTBEAT_TYPE,
            "sender_id": config.DEVICE_ID,
            "sender_name": config.DEVICE_NAME,
            "timestamp": time.time(),
        }).encode()

        # Multicast
        try:
            sock = socket.socket(
                socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP
            )
            sock.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_TTL, 2)
            sock.sendto(msg, (config.MULTICAST_GROUP, config.UDP_PORT))
            sock.close()
        except Exception as e:
            logger.debug(f"Multicast heartbeat failed: {e}")

        # Broadcast
        try:
            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
            sock.sendto(msg, ("255.255.255.255", config.UDP_PORT))
            sock.close()
        except Exception as e:
            logger.debug(f"Broadcast heartbeat failed: {e}")

    # --- Device Cleanup ---

    def _cleanup_loop(self):
        while self._running:
            time.sleep(5)
            now = time.time()
            with self._devices_lock:
                stale = [
                    did
                    for did, info in self._discovered_devices.items()
                    if now - info["last_seen"] > config.DEVICE_TIMEOUT_S
                ]
                for did in stale:
                    name = self._discovered_devices[did]["name"]
                    del self._discovered_devices[did]
                    logger.info(f"Device expired: {name} ({did})")

                if stale:
                    logger.info(
                        f"Nearby devices: {len(self._discovered_devices)}"
                    )

    # --- UDP Listener ---

    def _udp_listener(self):
        sock = None
        try:
            sock = socket.socket(
                socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP
            )
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

            # Platform-specific
            try:
                sock.setsockopt(
                    socket.SOL_SOCKET, socket.SO_REUSEPORT, 1
                )
            except AttributeError:
                pass  # Windows

            sock.bind(("", config.UDP_PORT))

            # Join multicast group
            try:
                group = socket.inet_aton(config.MULTICAST_GROUP)
                mreq = struct.pack("4sL", group, socket.INADDR_ANY)
                sock.setsockopt(
                    socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, mreq
                )
            except Exception as e:
                logger.warning(f"Multicast join failed: {e}")

            # Also enable broadcast receive
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)

            sock.settimeout(1.0)
            logger.info(f"UDP listener started on port {config.UDP_PORT}")

            while self._running:
                try:
                    data, addr = sock.recvfrom(4096)
                    msg = data.decode("utf-8")
                    self._handle_udp_message(msg, addr[0])
                except socket.timeout:
                    continue
                except Exception as e:
                    if self._running:
                        logger.error(f"UDP receive error: {e}")
                        time.sleep(1)

        except Exception as e:
            logger.error(f"UDP listener failed: {e}", exc_info=True)
        finally:
            if sock:
                sock.close()

    def _handle_udp_message(self, message: str, sender_ip: str):
        try:
            data = json.loads(message)
            msg_type = data.get("type")
            sender_id = data.get("sender_id")

            if sender_id == config.DEVICE_ID:
                return  # ignore own messages

            sender_name = data.get("sender_name", "Unknown")

            if msg_type == config.HEARTBEAT_TYPE:
                self._handle_heartbeat(sender_id, sender_name, sender_ip)

            elif msg_type == config.BROADCAST_TYPE_SCREENSHOT_READY:
                self._handle_heartbeat(sender_id, sender_name, sender_ip)

                offer = {
                    "sender_id": sender_id,
                    "sender_name": sender_name,
                    "sender_address": sender_ip,
                    "tcp_port": data["tcp_port"],
                    "file_size": data.get("file_size", 0),
                    "timestamp": data.get("timestamp", time.time()),
                }
                logger.info(
                    f"📡 Screenshot offer from {sender_name} "
                    f"({sender_ip}:{offer['tcp_port']})"
                )
                if self.on_screenshot_offer:
                    self.on_screenshot_offer(offer)

        except json.JSONDecodeError:
            logger.debug(f"Invalid JSON from {sender_ip}")
        except Exception as e:
            logger.error(f"Error handling UDP message: {e}")

    def _handle_heartbeat(
        self, sender_id: str, sender_name: str, sender_ip: str
    ):
        with self._devices_lock:
            is_new = sender_id not in self._discovered_devices
            self._discovered_devices[sender_id] = {
                "name": sender_name,
                "address": sender_ip,
                "last_seen": time.time(),
            }
            if is_new:
                count = len(self._discovered_devices)
                logger.info(
                    f"🔍 New device: {sender_name} ({sender_ip}) "
                    f"— total nearby: {count}"
                )

    # --- Screenshot Broadcast ---

    def broadcast_screenshot(self, screenshot_data: bytes):
        """Start TCP server and broadcast availability."""
        tcp_port = self._start_tcp_server(screenshot_data)

        msg = json.dumps({
            "type": config.BROADCAST_TYPE_SCREENSHOT_READY,
            "sender_id": config.DEVICE_ID,
            "sender_name": config.DEVICE_NAME,
            "tcp_port": tcp_port,
            "file_size": len(screenshot_data),
            "timestamp": time.time(),
        }).encode()

        # Multicast
        try:
            sock = socket.socket(
                socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP
            )
            sock.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_TTL, 2)
            sock.sendto(msg, (config.MULTICAST_GROUP, config.UDP_PORT))
            sock.close()
        except Exception as e:
            logger.debug(f"Multicast broadcast failed: {e}")

        # Broadcast
        try:
            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
            sock.sendto(msg, ("255.255.255.255", config.UDP_PORT))
            sock.close()
        except Exception as e:
            logger.debug(f"Broadcast failed: {e}")

        logger.info(f"Screenshot broadcast sent (port {tcp_port})")

    def _start_tcp_server(self, data: bytes) -> int:
        if self._tcp_server_socket:
            try:
                self._tcp_server_socket.close()
            except Exception:
                pass

        server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        server.bind(("", 0))
        port = server.getsockname()[1]
        server.listen(5)
        server.settimeout(config.SCREENSHOT_OFFER_TIMEOUT_S + 5)
        self._tcp_server_socket = server

        def serve():
            try:
                for _ in range(5):
                    try:
                        client, addr = server.accept()
                        threading.Thread(
                            target=self._handle_tcp_client,
                            args=(client, data, addr),
                            daemon=True,
                        ).start()
                    except socket.timeout:
                        break
            except Exception as e:
                if self._running:
                    logger.error(f"TCP server error: {e}")
            finally:
                server.close()

        threading.Thread(target=serve, daemon=True).start()
        logger.info(f"TCP server listening on port {port}")
        return port

    def _handle_tcp_client(
        self, client: socket.socket, data: bytes, addr
    ):
        try:
            request = client.recv(64).decode().strip()
            if request == "GET":
                header = struct.pack("!I", len(data))
                client.sendall(header + data)
                logger.info(
                    f"Sent {len(data)} bytes to {addr[0]}:{addr[1]}"
                )
        except Exception as e:
            logger.error(f"TCP client handler error: {e}")
        finally:
            client.close()

    # --- Download ---

    def download_screenshot(self, offer: dict) -> Optional[bytes]:
        try:
            addr = offer["sender_address"]
            port = offer["tcp_port"]

            logger.info(f"Connecting to {addr}:{port}...")
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.settimeout(10)
            sock.connect((addr, port))

            sock.sendall(b"GET\n")

            # Read 4-byte length header
            header = b""
            while len(header) < 4:
                chunk = sock.recv(4 - len(header))
                if not chunk:
                    raise ConnectionError("Connection closed reading header")
                header += chunk

            length = struct.unpack("!I", header)[0]
            logger.info(f"Receiving {length} bytes...")

            # Read data
            data = b""
            while len(data) < length:
                chunk = sock.recv(min(65536, length - len(data)))
                if not chunk:
                    raise ConnectionError("Connection closed reading data")
                data += chunk

            sock.close()
            logger.info(f"Download complete ({len(data)} bytes)")
            return data

        except Exception as e:
            logger.error(f"Download failed: {e}")
            return None

